File: ImperatorToCK3/Source/Mappers/SuccessionLawMapper/SuccessionLawMapper.cpp
#include "SuccessionLawMapper.h"
#include "SuccessionLawMapping.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



mappers::SuccessionLawMapper::SuccessionLawMapper() {
	LOG(LogLevel::Info) << "-> Parsing succession law mappings.";
	registerKeys();
	parseFile("configurables/succession_law_map.txt");
	clearRegisteredKeywords();
	LOG(LogLevel::Info) << "<> Loaded " << impToCK3SuccessionLawMap.size() << " succession law links.";
}


mappers::SuccessionLawMapper::SuccessionLawMapper(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void mappers::SuccessionLawMapper::registerKeys() {
	registerKeyword("link", [this](std::istream& theStream) {
		const SuccessionLawMapping mapping(theStream);
		if (mapping.ck3SuccessionLaws.empty())
			throw std::runtime_error("SuccessionLawMapper: link with no CK3 successions laws");
		
		auto [iterator, inserted] = impToCK3SuccessionLawMap.emplace(mapping.impLaw, mapping.ck3SuccessionLaws);
		if (!inserted) {
			iterator->second.insert(mapping.ck3SuccessionLaws.begin(), mapping.ck3SuccessionLaws.end());
		}
	});
	registerMatcher(commonItems::catchallRegexMatch, commonItems::ignoreItem);
}


std::set<std::string> mappers::SuccessionLawMapper::getCK3LawsForImperatorLaws(const std::set<std::string>& laws) const {
	std::set<std::string> lawsToReturn;
	for (const auto& impLaw : laws) {
		const auto& mapItr = impToCK3SuccessionLawMap.find(impLaw);
		if (mapItr != impToCK3SuccessionLawMap.end()) {
			lawsToReturn.insert(mapItr->second.begin(), mapItr->second.end());
		}
	}
	return lawsToReturn;
}

File: ImperatorToCK3/Mappers/SuccessionLaw/SuccessionLawMapper.cs
using commonItems;
using commonItems.Collections;
using ImperatorToCK3.CK3;
using System.Collections.Generic;

namespace ImperatorToCK3.Mappers.SuccessionLaw;

internal sealed class SuccessionLawMapper {
	private readonly List<SuccessionLawMapping> mappings = [];

	public SuccessionLawMapper() { }
	public SuccessionLawMapper(string filePath, OrderedDictionary<string, bool> ck3ModFlags) {
		Logger.Info("Parsing succession law mappings...");
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseLiquidFile(filePath, ck3ModFlags);
		Logger.Info($"Loaded {mappings.Count} succession law links.");

		Logger.IncrementProgress();
	}
	public SuccessionLawMapper(BufferedReader reader) {
		var parser = new Parser();
		RegisterKeys(parser);
		parser.ParseStream(reader);
	}
	private void RegisterKeys(Parser parser) {
		parser.RegisterKeyword("link", reader => {
			mappings.Add(new(reader));
		});
		parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
	}
	public OrderedSet<string> GetCK3LawsForImperatorLaws(SortedSet<string> impLaws, string? irGovernment, IReadOnlyCollection<string> enabledCK3Dlcs) {
		var lawsToReturn = new OrderedSet<string>();
		foreach (var impLaw in impLaws) {
			foreach (var mapping in mappings) {
				var match = mapping.Match(impLaw, irGovernment, enabledCK3Dlcs);
				if (match is null) {
					continue;
				}
				
				lawsToReturn.UnionWith(match);
				break;
			}
		}
		return lawsToReturn;
	}
}